### DISCUSSION TITLE
Let wifi-password change automatically

### DISCUSSION THREAD

**Discussion by vhDavid** [RaspAP/raspap-webgui#860] [Q&A]
Hi, I don't use RaspAP... yet... I am a volunteer in a visitor-center and am looking for wifi-hotspot software that has the ability to change the wifi-password automatically. The idea is to let our website generate a password at night with a chronjob and store it in a database. Next let the hotspot read it at boot (maybe a url in the config file instead of a password) and set it as wifi password and let another pi also read it at boot and display it on a screen. Another option is that the hotspot software generates a password at boot and let the other pi read it somehow at boot and display it. The first option would be the most secure because they guaranteed have the same password. Is that possible with RaspAP in any way? Maybe a mod I can do myself?

**Comment by billz** [comment]
Interesting question. Here's [one way to do it](https://gist.github.com/billz/2cc43e96563293d650e313e068d52dfb) with bash.

This generates a strong random password, updates hostapd and restarts RaspAP. 
The new passphrase is visible on the **Hotspot > Security** tab and also via the QR code.

**Reply by vhDavid** [reply]
Thanks Billz, I will try that. I want it to be readable by another pi tho. Maybe I can adapt this script to let it visit a url to get that php-generated password.

**Reply by billz** [reply]
Assuming both Pi's are reachable on the same network, you could use `scp` to share hostapd.conf with the second Pi. There are other ways of sharing a passphrase, but this is more secure than passing it via a URL (scp provides the same security as ssh). If you only need to display the AP login, you could put the second Pi into [monitor mode](https://docs.raspap.com/faq/#monitor) or adapt the script to output what you need as static html.

**Reply by vhDavid** [reply]
Thanks. I was able to install RaspAP and also your script does it's job. But the password needs to be displayed on a second pi on a page where also other info is displayed. I tried to get the password from our website by adapting your script with 'curl' and 'wget' but it does not work due to a dns issue. Probably because it is handled by dnsmasq. You write to use scp, but then the problem remains that - when the password-generating pi restarts - the displaying pi needs to reload or it displays the previous password. There's also a server (synology) in our building, I could try to get the password there (no dns needed) and change the password with a python script. If there is another solution to get it from our website without disabling the adblocker please do tell.

**Reply by billz** [reply]
> when the password-generating pi restarts - the displaying pi needs to reload or it displays the previous password

Can you ssh between the Pi's?  Copy your ssh keys to the displaying pi with `ssh-copy-id`. Then extend the script to generate some html and `ssh` it over to the displaying pi:

```
msg="<html><body>The new passphrase is: ${PASSPHRASE}</body></html>"
echo "${msg}" | ssh pi@192.168.1.8 'cat > notice.html'
```

**Reply by vhDavid** [reply]
I will try that when I find another pi to test at home. That pi is in our visitor center now only displaying a website with lots of info. Still hope an option like this gets implemented in the software. Thanks for all the info.

**Reply by vhDavid** [reply]
Hi Billz, just to let you know I managed to get it done by adapting your script and using the synology nas. Not by ssh because it seems only possible to ssh into that as root user. I now change the password with python and let both pi's read it on reboot from a map only shared to those pi's. Thanks again.

**Reply by billz** [reply]
Glad to hear it. Have you considered documenting your setup? I think others could benefit from it.

**Reply by vhDavid** [reply]
Hope I did not forget anything and placing it here is ok... I'm pretty new here.

This is the used python script: People can choose which characters they want to include.

```
from random import randrange

karakters = "abcdefghijkmnpqrstuvwxyzABCDEFGHJKLMNPQRSTUVWXYZ23456789"
lengte = len(karakters)
paswoord = ""

for x in range(6):
    welk = randrange(lengte)
    paswoord += karakters[welk:welk+1]

f = open("/volume1/RaspberryPi/wifi_paswoord.txt", "w")
f.write(paswoord)
f.close()
```
The nas needs to support python. I schedule the scipt as root every day at 3AM.
The shared map 'RaspberryPi' has read-rights for the users for the 2 pi's.

In both pi's I added this line to the etc/fstab to be able to access the map on the nas:
//192.168.1.134/xxxx	/home/pi/yyyy	cifs	credentials=/home/pi/.zzzz,iocharset=utf8,users,_netdev	0	0

The 2nd part is a map on the pi you need to make yourself where the remote map gets 'mapped' when it is mounted.

The credentials-file (.zzzz) contains following lines:
username=
password=

Then there is Billz script a bit adapted that is scheduled on boot with crontab.

```
#!/bin/bash

readonly HOSTAPD="/etc/hostapd/hostapd.conf"
readonly PASSPHRASE=$(cat "/home/pi/yyyy/paswoord.txt")
echo "Your new passphrase is: ${PASSPHRASE}"

# update hostapd.conf wpa_passphrase
sudo sed -i "s/\(wpa_passphrase=\).*/\1${PASSPHRASE}/g" "$HOSTAPD" || { echo  "Unable to modify hostapd.conf" && exit;$

# restart RaspAP for change to take effect
sudo systemctl restart raspapd.service || { echo  "Unable to restart raspapd.service" && exit; }
```
On the 2nd pi 'PiPresents' is installed. It is a presentation-software for the raspberry pi that can display info from different sources together on 1 tv screen. [More info on PiPresents](https://pipresents.wordpress.com/)

I noticed that 2nd pi needed an extra mount-command for the nas in the script that also auto-starts pipresents.
sudo mount –a
